### PR TITLE
task4

### PR DESCRIPTION
**Pull request by ArtemPrydorozhko**
# shop-services
task4.1 - done (database instance in RDS is created, SQL scripts are present) \
task4.2 - done (lambda functions are integrated with DB) \
task4.3 - done (createProduct is present and working) 

FE does not need any changes (product model is correct, interaction with AWS lambda was done in previous task), link to repo https://github.com/ArtemPrydorozhko/shop-angular-cloudfront

get product by id https://kzdx8lc0v0.execute-api.eu-west-1.amazonaws.com/dev/products/67efc835-a45b-40d1-b107-77d802111651

FE app
All products http://testshopawsbucket2.s3-website-us-east-1.amazonaws.com \
add a new product link http://testshopawsbucket2.s3-website-us-east-1.amazonaws.com/admin/products/new

Additional (optional) tasks
 - POST /products lambda functions return error 400 status code if product data is invalid
 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
 - All lambdas do console.log for each incoming requests
 - Transaction based creation of product ("BEGIN", "COMMIT", "ROLLBACK" queries in createProduct.js)

### CODE REVIEW

**Comment by jafrias2159** [comment]
Looks good to me!

File: product-service/handlers/getProductsList.js [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
-  };
+  try {
+    console.log("getProductsList", JSON.stringify(event));
+
+    await client.connect();
+
+    const { rows: products } = await client.query(
+      "SELECT * from products JOIN stocks ON products.id = stocks.product_id"
+    );
+    await client.end();
```
Comment (siargei-stepanov) [inline product-service/handlers/getProductsList.js:16]: `client.end();` can be moved to `finally` block

File: product-service/handlers/getProductsList.js [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
-    headers,
-    body: JSON.stringify(mockProducts),
-  };
+  try {
+    console.log("getProductsList", JSON.stringify(event));
+
+    await client.connect();
+
+    const { rows: products } = await client.query(
+      "SELECT * from products JOIN stocks ON products.id = stocks.product_id"
```
Comment (siargei-stepanov) [inline product-service/handlers/getProductsList.js:14]: You can select only the fileds you really need. If you check current response, you'll see 2 identical fields: `id` and `product_id`

File: product-service/utils/productValidator.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export default (product) => {
+  return (
+    typeof product.title === "string" &&
```
Comment (siargei-stepanov) [inline product-service/utils/productValidator.js:3]: You are not checking the values to be not empty. It means you can create a product with title = '"".

**Review by siargei-stepanov — APPROVED** [review]
Looks good, but can be improved

File: product-service/utils/productValidator.js (hunk 3 above)
Comment (ArtemPrydorozhko) [reply product-service/utils/productValidator.js:3]: Thanks for the comments! I will update them in the next PR.

**Pull request merged**